pkg/querier/parquet_queryable.go

```diff
@@ -643,35 +652,81 @@ func newCacheMetrics(reg prometheus.Registerer) *cacheMetrics {
 	}
 }
 
+type cacheEntry[T any] struct {
+	value     T
+	expiresAt time.Time
+}
+
 type Cache[T any] struct {
-	cache   *lru.Cache[string, T]
+	cache   *lru.Cache[string, *cacheEntry[T]]
 	name    string
 	metrics *cacheMetrics
+	ttl     time.Duration
+	stopCh  chan struct{}
 }
 
-func newCache[T any](name string, size int, metrics *cacheMetrics) (cacheInterface[T], error) {
+func newCache[T any](name string, size int, ttl, maintenanceInterval time.Duration, metrics *cacheMetrics) (cacheInterface[T], error) {
 	if size <= 0 {
 		return &noopCache[T]{}, nil
 	}
-	cache, err := lru.NewWithEvict(size, func(key string, value T) {
+	cache, err := lru.NewWithEvict(size, func(key string, value *cacheEntry[T]) {
 		metrics.evictions.WithLabelValues(name).Inc()
 		metrics.size.WithLabelValues(name).Dec()
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	return &Cache[T]{
+	c := &Cache[T]{
 		cache:   cache,
 		name:    name,
 		metrics: metrics,
-	}, nil
+		ttl:     ttl,
+		stopCh:  make(chan struct{}),
+	}
+
+	if ttl > 0 {
+		go c.maintenanceLoop(maintenanceInterval)
+	}
+
+	return c, nil
+}
+
+func (c *Cache[T]) maintenanceLoop(interval time.Duration) {
```

Comment:
Just a note, I think we need to move this cache implementation or some reusable parquet code into a separate package so that store gateway parquet mode can reuse the same thing